### PR TITLE
Add memory targetAverageUtilization config

### LINKED ISSUE #47649: Add Memory Based Autoscaling on Istioctl Install
**Describe the feature request**

Since Kubernetes now supports Memory based Autoscaling for HPA, it would be great to add a feature like that to the operator (even though it's not prioritized anymore). I will try to contribute myself if I have enough time 

```
cpu:
  targetAverageUtilization: 80
memory: #ADD THIS
  targetAverageUtilization: 80 #ADD THIS
```

Complete example

```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  components:
    base:
      enabled: true
    cni:
      enabled: false
    egressGateways:
    - enabled: false
      name: istio-egressgateway
    ingressGateways:
    - enabled: true
      name: istio-ingressgateway
    istiodRemote:
      enabled: false
    pilot:
      enabled: true
  hub: docker.io/istio
  meshConfig:
    defaultConfig:
      proxyMetadata: {}
    enablePrometheusMerge: true
  profile: default
  tag: 1.19.1
  values:
    base:
      enableCRDTemplates: false
      validationURL: ""
    defaultRevision: ""
    gateways:
      istio-egressgateway:
        autoscaleEnabled: true
        env: {}
        name: istio-egressgateway
        secretVolumes:
        - mountPath: /etc/istio/egressgateway-certs
          name: egressgateway-certs
          secretName: istio-egressgateway-certs
        - mountPath: /etc/istio/egressgateway-ca-certs
          name: egressgateway-ca-certs
          secretName: istio-egressgateway-ca-certs
        type: ClusterIP
      istio-ingressgateway:
        autoscaleEnabled: true
        env: {}
        name: istio-ingressgateway
        secretVolumes:
        - mountPath: /etc/istio/ingressgateway-certs
          name: ingressgateway-certs
          secretName: istio-ingressgateway-certs
        - mountPath: /etc/istio/ingressgateway-ca-certs
          name: ingressgateway-ca-certs
          secretName: istio-ingressgateway-ca-certs
        type: LoadBalancer
    global:
      configValidation: true
      defaultNodeSelector: {}
      defaultPodDisruptionBudget:
        enabled: true
      defaultResources:
        requests:
          cpu: 10m
      imagePullPolicy: ""
      imagePullSecrets: []
      istioNamespace: istio-system
      istiod:
        enableAnalysis: false
      jwtPolicy: third-party-jwt
      logAsJson: false
      logging:
        level: default:info
      meshNetworks: {}
      mountMtlsCerts: false
      multiCluster:
        clusterName: ""
        enabled: false
      network: ""
      omitSidecarInjectorConfigMap: false
      oneNamespace: false
      operatorManageWebhooks: false
      pilotCertProvider: istiod
      priorityClassName: ""
      proxy:
        autoInject: enabled
        clusterDomain: cluster.local
        componentLogLevel: misc:error
        enableCoreDump: false
        excludeIPRanges: ""
        excludeInboundPorts: ""
        excludeOutboundPorts: ""
        image: proxyv2
        includeIPRanges: '*'
        logLevel: warning
        privileged: false
        readinessFailureThreshold: 30
        readinessInitialDelaySeconds: 1
        readinessPeriodSeconds: 2
        resources:
          limits:
            cpu: 2000m
            memory: 1024Mi
          requests:
            cpu: 100m
            memory: 128Mi
        statusPort: 15020
        tracer: zipkin
      proxy_init:
        image: proxyv2
      sds:
        token:
          aud: istio-ca
      sts:
        servicePort: 0
      tracer:
        datadog: {}
        lightstep: {}
        stackdriver: {}
        zipkin: {}
      useMCP: false
    istiodRemote:
      injectionURL: ""
    pilot:
      autoscaleEnabled: true
      autoscaleMax: 5
      autoscaleMin: 1
      configMap: true
      cpu:
        targetAverageUtilization: 80
      memory: #ADD THIS
        targetAverageUtilization: 80 #ADD THIS
      env: {}
      image: pilot
      keepaliveMaxServerConnectionAge: 30m
      nodeSelector: {}
      podLabels: {}
      replicaCount: 1
      traceSampling: 1
    telemetry:
      enabled: true
      v2:
        enabled: true
        metadataExchange:
          wasmEnabled: false
        prometheus:
          enabled: true
          wasmEnabled: false
        stackdriver:
          configOverride: {}
          enabled: false
          logging: false
          monitoring: false
          topology: false
```

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

### PR DESCRIPTION
**Pull request by rlaisqls**
**Please provide a description of this PR:**

Added settings, fields, and tests for memory-based automatic scaling for HPA to enable more detailed settings for the istio resource HPA.

resolved #47649

### CODE REVIEW

File: manifests/charts/gateways/istio-egress/values.yaml [hunk 1]
```diff
@@ -33,6 +33,8 @@ gateways:
         memory: 1024Mi
     cpu:
       targetAverageUtilization: 80
+    memory:
+      targetAverageUtilization: 80
```
Comment (howardjohn) [inline manifests/charts/gateways/istio-egress/values.yaml:37]: This makes it _required_ to set a memory target. This should be opt in.

**Review by howardjohn — CHANGES_REQUESTED** [review]
IMO scaling on memory is not very useful. @briansonnenberg can probably provide more context.

I don't mind making it configurable but it must be off by default

File: manifests/charts/gateways/istio-egress/values.yaml (hunk 1 above)
Comment (rlaisqls) [reply manifests/charts/gateways/istio-egress/values.yaml:37]: Okay, I'll modify it to apply the option selectively.

**Comment by rlaisqls** [comment]
I found a problem with the test code and behavior, but I kept the PR closed because I didn't have enough time to work. I will review the code again and open it soon. I'm sorry :(

**Comment by samuel-esp** [comment]
> IMO scaling on memory is not very useful. @briansonnenberg can probably provide more context.
> 
> I don't mind making it configurable but it must be off by default

@howardjohn @briansonnenberg can you provide more context to this? We run many production clusters with thousands of pods (between 500 to 3000, most clusters have around 1000/1500 pods). We noticed a good improvement on istio performances by scaling on memory, especially with Istiod. To me it seems that scaling on memory also help during a cpu spike. If autoscaling based on memory is not available the cluster operator would need to fine tune Istiod/IngressGateway memory almost day to day using a dediated VPA object, this can be tedious especially if the cluster is continuously onboarding new workloads 

I see the PR is closed, if @rlaisqls is busy I'll try to make a new one as soon as I have more time outside work

**Comment by rlaisqls** [comment]
@howardjohn I added memory targetAverageUtilization options to selectively apply when creating igress, egress, and istiod, and corrected test errors due to data snapshots. 
@samuel-esp If you don't mind, please review if this PR fits your requirements.

**Comment by howardjohn** [comment]
> > IMO scaling on memory is not very useful. @briansonnenberg can probably provide more context.
> > I don't mind making it configurable but it must be off by default
> 
> @howardjohn @briansonnenberg can you provide more context to this? We run many production clusters with thousands of pods (between 500 to 3000, most clusters have around 1000/1500 pods). We noticed a good improvement on istio performances by scaling on memory, especially with Istiod. To me it seems that scaling on memory also help during a cpu spike. If autoscaling based on memory is not available the cluster operator would need to fine tune Istiod/IngressGateway memory almost day to day using a dedicated VPA object or other monitoring tools, this can be tedious especially if the cluster is continuously onboarding new workloads
> 
> I see the PR is closed, if @rlaisqls is busy I'll try to make a new one as soon as I have more time outside work

Problem is, unlike CPU (ish), memory does not scale only with load. There is a baseline memory usage and a per-load memory usage.

For istiod, its very much baseline-memroy heavy. Istiod with 1 client vs 1000 clients will probably use nearly the same memory.

Envoy scales a bit more with request load, but still depends on the environment. 

If each workload uses a constant baseline memory, and we scale up based on memory, the results are terrible -- we will scale forever and not solve any problems.

Absolutely there are use cases where it is useful, but I don't think its always useful and  not safe as a default.

File: manifests/charts/gateway/values.yaml [hunk 2]
```diff
@@ -73,6 +73,7 @@ autoscaling:
   minReplicas: 1
   maxReplicas: 5
   targetCPUUtilizationPercentage: 80
+  targetMemoryUtilizationPercentage: 80
```
Comment (howardjohn) [inline manifests/charts/gateway/values.yaml:76]: I think you missed turning this one off by default?

Comment (rlaisqls) [reply manifests/charts/gateway/values.yaml]: I think I missed one part... Thank you. I'll fix it :)

File: pkg/config/analysis/analyzers/testdata/common/sidecar-injector-configmap.yaml [hunk 3]
```diff
@@ -406,7 +406,7 @@ data:
...
...ilable":"25%","selfSigned":false,"tolerations":[]},"telemetry":{"enabled":true,"v1":{"enabled":true},"v2":{"enabled":false,"prometheus":{"enabled":true},"stackdriver":{"configOverride":{},"enabled":false,"logging":false,"monitoring":false,"topology":false}}},"tracing":{"enabled":false,"ingress":{"enabled":false},"jaeger":{"accessMode":"ReadWriteMany","enabled":false,"hub":"docker.io/jaegertracing","memory":{"max_traces":50000},"namespace":"istio-system","persist":false,"spanStorageType":"badger","storageClassName":"","tag":"1.20"},"nodeSelector":{},"opencensus":{"exporters":{"stackdriver":{"enable_tracing":true}},"hub":"docker.io/omnition","resources":{"limits":{"cpu":"1","memory":"2Gi"},"requests":{"cpu":"200m","memory":"400Mi"}},"tag":"0.1.9"},"podAntiAffinityLabelSelector":[],"podAntiAffinityTermLabelSelector":[],"provider":"jaeger","service":{"annotations":{},"externalPort":9411,"name":"http-query","type":"ClusterIP"},"zipkin":{"hub":"docker.io/openzipkin","javaOptsHeap":700,"maxSpans":500000,"node":{"cpus":2},"probeStartupDelay":200,"queryPort":9411,"resources":{"limits":{"cpu":"300m","memory":"900Mi"},"requests":{"cpu":"150m","memory":"900Mi"}},"tag":"2.14.2"}},"version":""}'
```
Comment (hanxiaop) [inline pkg/config/analysis/analyzers/testdata/common/sidecar-injector-configmap.yaml:409]: It seems like you've updated the input files for analysis, but no test cases for the analysis have been updated. Maybe just revert this change. Will it affect the tests?

File: manifests/charts/gateways/istio-ingress/values.yaml [hunk 4]
```diff
@@ -35,6 +35,8 @@ gateways:
 
     cpu:
       targetAverageUtilization: 80
+    memory: {}
```
Comment (hanxiaop) [inline manifests/charts/gateways/istio-ingress/values.yaml:38]: Can we only add this to the `manifests/charts/gateway`? It's not recommended to use istio-ingress and istio-egress anymore for a long time. See https://github.com/istio/istio/tree/master/manifests/charts/gateway#legacy-gateway-helm-charts

File: pkg/config/analysis/analyzers/testdata/common/sidecar-injector-configmap.yaml (hunk 3 above)
Comment (rlaisqls) [reply pkg/config/analysis/analyzers/testdata/common/sidecar-injector-configmap.yaml:409]: I don't think the existing cpu targetAverageUtilization value also affect in the file. 
However, since the value for memory is not the default value, it is correct that it should be further deleted. I will correct it.

File: manifests/charts/gateways/istio-ingress/values.yaml (hunk 4 above)
Comment (rlaisqls) [reply manifests/charts/gateways/istio-ingress/values.yaml:38]: I didn't know that. It's better to erase it. Thank you :)

**Review by howardjohn — APPROVED** [review]

**Review by hanxiaop — APPROVED** [review]
LGTM
